packages/lib-classifier/src/plugins/drawingTools/experimental/models/marks/FreehandLine/FreehandLine.js

```diff
@@ -3,278 +3,603 @@ import { FreehandLine as FreehandLineComponent } from '@plugins/drawingTools/com
-      const clippedPoints = self.points.slice(firstPoint, lastPoint + 1)
-      self.clipPath = clippedPoints.map(({ x, y }) => ({ x, y }))
-      self.pathX.splice(firstPoint + 1, deleteCount)
-      self.pathY.splice(firstPoint + 1, deleteCount)
-      // Make the ends of the spliced section draggable.
-      self.dragPoint = self.points[firstPoint]
-      self.targetPoint = self.points[firstPoint + 1]
-    },
 
-    trim(startIndex, endIndex) {
-      let firstPoint = Math.min(startIndex, endIndex)
-      let lastPoint = Math.max(startIndex, endIndex)
-      /*
-        Reverse the path, if necessary, so that the first point clicked
-        is always at the end of the new line. The end of an open line
-        is the draggable point.
-      */
-      if (lastPoint === endIndex) {
-        const lastIndex = self.points.length - 1
-        self.pathX.reverse()
-        self.pathY.reverse()
-        lastPoint = lastIndex - startIndex
-        firstPoint = lastIndex - endIndex
+      if (spliceReverse && dragPointIndex < spliceSpan) {
+        spliceThroughBeginning = true;
+      } else if (!spliceReverse && (dragPointIndex + spliceSpan) > self.points.length) {
+        spliceThroughBeginning = true;
       }
-      const clippedPoints = [ ...self.points.slice(lastPoint), ...self.points.slice(0, firstPoint + 1)]
-      self.clipPath = clippedPoints.map(({ x, y }) => ({ x, y }))
-      // Move the end of the line back to lastPoint.
-      self.pathX.splice(lastPoint + 1)
-      self.pathY.splice(lastPoint + 1)
-      // Move the start of the line forward to firstPoint.
-      self.pathX.splice(0, firstPoint)
-      self.pathY.splice(0, firstPoint)
-      // Make the ends of the new, open line draggable.
-      self.dragPoint = null
-      self.targetPoint = null
+
+      self.drawActions.push({
+        type: 'splice-close-point',
+        pointIndexDrag: dragPointIndex,
+        pointIndexClose: closePointIndex,
+        pointsFromDrag: [],
+        pointsFromClose: [],
+        pointsSpliceCount: 0,
+        isReversed: spliceReverse,
+        throughBeginning: spliceThroughBeginning
+      });
+
+      self.splicePathSetup()
+      self.redoActionsClear()
+    },
+
+    splicePathSetup() {
+      // for purposes of undo/redo we use the action to set these properties
+      let action = self.drawActions.at(-1)
+      self.spliceClosePointIndex = action.pointIndexClose
+      self.spliceReverse = action.isReversed
+      self.spliceThroughBeginning = action.throughBeginning
+      self.spliceActive = true
+      self.setClosePoint(self.points[action.pointIndexClose])
+    },
+
+    spliceReset() {
+      self.splicePoints.length = 0
+      self.spliceActive = false
+      self.spliceReverse = false
+      self.spliceThroughBeginning = false
+      self.spliceDragPointIndex = null
+      self.spliceClosePointIndex = null
+    },
+
+    move() {
+      // do nothing... tool required but not used in this context
+    },
+
+    finish({ x, y }) {
```

Comment:
This errors when I press Done. I think that's because `mark.finish()` doesn't take arguments in the drawing API.

Reply:
It looks like x and y aren't used anyway, so that's easy to fix.

Reply:
If I fix this bug locally, then I can post classifications to Panoptes staging. However, the annotations are empty because the `FreehandLine` model is empty. The model's `pathX` and `pathY` properties should be sent back with the classification.

<img width="827" alt="Screenshot of a freehand line classification in browser dev tools, showing that the mark annotation is empty." src="https://user-images.githubusercontent.com/59547/224003434-3fbf7bc5-c4c5-462c-ab63-602d573509c3.png">

Reply:
Good catch - fixed